File: demo/py/bfx.wss11.py

import os
import sys
import logging

import json

import threading

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../code')))

#from adapters import AdpBitfinexWSS
from ktdata import CTNetClient_BfxWss
from ktdata import CTDataSet_Ticker_DbIn, CTDataSet_Ticker_DbOut, CTDataSet_ABooks_DbIn, CTDataSet_ABooks_DbOut, CTDataSet_ACandles_DbIn, CTDataSet_ACandles_DbOut
from ktdata import KTDataMedia_DbReader, KTDataMedia_DbWriter

from pymongo import MongoClient

API_KEY     = None
API_SECRET  = None

obj_netclient = None
obj_dbreader  = None
obj_dbwriter = None

flag_netclient = False

test_dataset_ticker = None
test_dataset_book = None
test_dataset_candles = None

logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
logger = logging.getLogger()

timer_count = 0

def timer_entr():
	global timer_count
	timer_count += 1
	logger.info("Timer: count=" + str(timer_count))
	if (obj_netclient != None) and (timer_count > 100):
		obj_netclient.close()
	else:
		threading.Timer(2.0, timer_entr).start()

str_db_uri  = 'mongodb://127.0.0.1:27017'
str_db_name = 'bfx-pub'

#flag_netclient = True

mapWREQs = [
		{ 'channel':  'ticker', 'uid': 'container-ticker', 'visible': False, 'wreq_args': { 'symbol': 'tBTCUSD', }, },
		{ 'channel':    'book', 'uid': 'container-bookP0', 'visible': False, 'wreq_args': { 'symbol': 'tBTCUSD', 'prec': 'P0', 'freq': 'F0', 'len': 100, }, },
		{ 'channel':    'book', 'uid': 'container-bookP1', 'visible': False, 'wreq_args': { 'symbol': 'tBTCUSD', 'prec': 'P1', 'freq': 'F0', 'len': 100, }, },
		{ 'channel': 'candles', 'uid': 'container-candle', 'visible':  True, 'wreq_args': { 'key': 'trade:1m:tBTCUSD', }, },
	]

if flag_netclient:
	url_bfx  = "wss://api.bitfinex.com/ws/2"
	obj_netclient = CTNetClient_BfxWss(logger, url_bfx)
	obj_dbwriter  = KTDataMedia_DbWriter(logger)
	obj_dbwriter.dbOP_Connect(str_db_uri, str_db_name)
else:
	obj_dbreader  = KTDataMedia_DbReader(logger)
	obj_dbreader.dbOP_Connect(str_db_uri, str_db_name)

for map_idx, map_unit in enumerate(mapWREQs):
	if not map_unit['visible']:
		continue
	obj_chan  = None
	logger.info("map idx=" + str(map_idx) + ", unit=" + str(map_unit))

	if map_unit['channel'] == 'ticker':
		if not flag_netclient:
			chan_obj = CTDataSet_Ticker_DbIn(logger, obj_dbreader, map_unit['wreq_args'])
		else:
			chan_obj = CTDataSet_Ticker_DbOut(logger, obj_dbwriter, map_unit['wreq_args'])
			test_dataset_ticker = chan_obj
	elif map_unit['channel'] == 'book':
		if not flag_netclient:
			chan_obj = CTDataSet_ABooks_DbIn(logger, obj_dbreader, map_unit['wreq_args'])
			test_dataset_book = chan_obj
		else:
			chan_obj = CTDataSet_ABooks_DbOut(logger, obj_dbwriter, map_unit['wreq_args'])
	elif map_unit['channel'] == 'candles':
		if not flag_netclient:
			chan_obj = CTDataSet_ACandles_DbIn(logger, obj_dbreader, 1000, map_unit['wreq_args'])
			test_dataset_candles = chan_obj
		else:
			chan_obj = CTDataSet_ACandles_DbOut(logger, obj_dbwriter, 1000, map_unit['wreq_args'])

	if chan_obj != None and obj_netclient != None:
		obj_netclient.addObj_DataReceiver(chan_obj)

timer_entr()

if flag_netclient:
	obj_netclient.run_forever()
elif test_dataset_ticker != None:
	#coll_name = "ticker-201712060400"
	coll_name = "ticker-201712060500"
	obj_dbreader.dbOP_LoadColl(coll_name, test_dataset_ticker, {}, [('mts', 1)], True)
elif test_dataset_book != None:
	coll_name = "book-P1-201712060330"
	obj_dbreader.dbOP_LoadColl(coll_name, test_dataset_book, {}, [('mts', 1)], True)
elif test_dataset_candles != None:
	coll_name = "candles-1m-201712060530"
	obj_dbreader.dbOP_LoadColl(coll_name, test_dataset_candles, {}, [('mts', 1)], True)

File: demo/js/bfx.wss11.js

const dev31 = require('../../code/js/kkai-dev31.js');

const dev51 = require('../../code/js/kkai-dev51.js');
const dev52 = require('../../code/js/kkai-dev52.js');

var flag_netclient  = true;
//flag_netclient  = false;

var obj_netclient = null;
var obj_dbreader  = null;
var obj_dbwritter = null;


var mapWREQs = [
        { channel:  'ticker', uid: 'container-ticker', visible:  true, wreq_args: { symbol: 'tBTCUSD', }, },
        { channel:    'book', uid: 'container-bookP0', visible:  true, wreq_args: { symbol: 'tBTCUSD', prec: 'P0', freq: 'F0', len: 100, }, },
        { channel:    'book', uid: 'container-bookP1', visible: false, wreq_args: { symbol: 'tBTCUSD', prec: 'P1', freq: 'F0', len: 100, }, },
        { channel: 'candles', uid: 'container-candle', visible: false, wreq_args: { key: 'trade:1m:tBTCUSD', }, },
      ];

var db_url_p  = "mongodb://localhost:27017";
var db_name   = "bfx-pub";

process.on('SIGINT', function() {
  console.log("WARN: caught interrupt signal ...");
/*
  if (obj_dbwritter != null) {
    obj_dbwritter.dbOP_Close();
  }
  else {
    process.exit();
  }
// */
    process.exit();
});

console.log("main(1) ...");

if (!flag_netclient) {
  obj_dbreader  = new dev52.ClDataSet_DbReader();
  obj_dbreader.dbOP_Connect(db_url_p + '/' + db_name);
}
else {
  obj_dbwritter = new dev52.ClDataSet_DbWriter();
  obj_dbwritter.dbOP_Connect(db_url_p + '/' + db_name);
}

if (flag_netclient) {
  obj_netclient = new dev31.ClNetClient_BfxWss();
}

var test_dataset_book = null;

for (var mi=0; mi < mapWREQs.length; mi++)
{
  var chan_obj;
  var map_unit = mapWREQs[mi];
  if (!map_unit.visible) {
    continue;
  }
  chan_obj = null;
  if (map_unit.channel == 'ticker') {
    if (!flag_netclient) {
      chan_obj = new dev51.ClDataSet_Ticker_DbIn(obj_dbreader, map_unit.wreq_args);
    }
    else {
      chan_obj = new dev51.ClDataSet_Ticker_DbOut(obj_dbwritter, map_unit.wreq_args);
    }
  }
  else
  if (map_unit.channel == 'book') {
    if (!flag_netclient) {
      chan_obj = new dev51.ClDataSet_ABooks_DbIn(obj_dbwritter, map_unit.wreq_args);
test_dataset_book = chan_obj;
    }
    else {
      chan_obj = new dev51.ClDataSet_ABooks_DbOut(obj_dbwritter, map_unit.wreq_args);
    }
  }
  else
  if (map_unit.channel == 'candles') {
    if (!flag_netclient) {
      chan_obj = new dev51.ClDataSet_ACandles_DbIn(1000, obj_dbwritter, map_unit.wreq_args);
    }
    else {
      chan_obj = new dev51.ClDataSet_ACandles_DbOut(1000, obj_dbwritter, map_unit.wreq_args);
    }
  }

  if ((chan_obj != null) && (obj_netclient != null)) {
    obj_netclient.addObj_DataReceiver(chan_obj);
  }
}

if (!flag_netclient) {
  obj_dbreader.dbOP_LoadColl('book-P0-201712011100', test_dataset_book);
}
else {
  obj_netclient.ncOP_Exec();
}

